File: Claude/Simples/graph/graph.cpp
// How to implement a graph? in c++

#include <iostream>
#include <vector>
#include <list>

// Graph class
class Graph {
private:
    int numVertices;
    std::vector<std::list<int>> adjacencyList;

public:
    // Constructor
    Graph(int vertices) {
        numVertices = vertices;
        adjacencyList.resize(vertices);
    }

    // Add an edge
    void addEdge(int src, int dest) {
        adjacencyList[src].push_back(dest);
        adjacencyList[dest].push_back(src); // Undirected graph
    }

    // Print the graph
    void printGraph() {
        for (int i = 0; i < numVertices; ++i) {
            std::cout << "Vertex " << i << ":";
            for (int neighbor : adjacencyList[i]) {
                std::cout << " -> " << neighbor;
            }
            std::cout << std::endl;
        }
    }
};

int main() {
    Graph g(3);

    g.addEdge(0, 1);
    g.addEdge(1, 0);
    g.addEdge(2, 2);

    g.printGraph();

    return 0;
}

/*
Output:
    Vertex 0: -> 1 -> 1
    Vertex 1: -> 0 -> 0
    Vertex 2: -> 2 -> 2
    Wrong:  Vertex 2: -> 2 -> 2
    Should be:  Vertex 2: -> 2
    Score: 0.75
*/

File: Claude/Simples/graph/graph.py
# Input: How to implement a graph? in python

class Graph:
    def __init__(self):
        self.adjacency_list = {}

    def add_vertex(self, vertex):
        if vertex not in self.adjacency_list:
            self.adjacency_list[vertex] = []

    def add_edge(self, vertex1, vertex2):
        if vertex1 in self.adjacency_list and vertex2 in self.adjacency_list:
            self.adjacency_list[vertex1].append(vertex2)
            self.adjacency_list[vertex2].append(vertex1)

    def print_graph(self):
        for vertex in self.adjacency_list:
            print(vertex, ":", self.adjacency_list[vertex])

# Example usage
graph = Graph()
graph.add_vertex('A')
graph.add_vertex('B')
graph.add_vertex('C')
graph.add_edge('A', 'B')
graph.add_edge('B', 'A')
graph.add_edge('C', 'C')
graph.print_graph()

"""
Output:
    A : ['B', 'B']
    B : ['A', 'A']
    C : ['C', 'C']

Wrong: C : ['C', 'C']
Should be: C : ['C']
Score: 0.75
"""